### ISSUE TITLE
KokkosKernels:  support for TSQR

### ISSUE THREAD

**Issue by kddevin** [kokkos/kokkos-kernels#567]
The TSQR solver in Trilinos currently calls cusolver directly.  We're rather call KokkosKernels.  Here are the cusolver functions that we call:
```
cusolverDnDgeqrf_bufferSize 
cusolverDnDgeqrf 
cusolverDnDormqr_bufferSize 
cusolverDnDormqr 
cusolverDnDorgqr_bufferSize
cusolverDnDorgqr
cusolverDnSgeqrf_bufferSize 
cusolverDnSgeqrf 
cusolverDnSormqr_bufferSize 
cusolverDnSormqr 
cusolverDnSorgqr_bufferSize
cusolverDnSorgqr
cusolverDnZgeqrf_bufferSize 
cusolverDnZgeqrf 
cusolverDnZunmqr_bufferSize 
cusolverDnZunmqr 
cusolverDnZungqr_bufferSize
cusolverDnZungqr
cusolverDnCgeqrf_bufferSize 
cusolverDnCgeqrf 
cusolverDnCunmqr_bufferSize 
cusolverDnCunmqr 
cusolverDnCungqr_bufferSize
cusolverDnCungqr
```

The use of these functions can be seen in 
https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/tsqr/src/Tsqr_Impl_CuSolver.cpp
Here you can see an abstraction to six functions, templated on scalar type:
```
    apply_Q_factor, apply_Q_factor_lwork
    compute_QR, compute_QR_lwork
    compute_explicit_Q. compute_explicit_Q_lwork
```
Maybe these functions would be the right level of abstraction.  Alternatively, higher levels (such as the NodeTsqr class https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/tsqr/src/Tsqr_NodeTsqr.hpp ) might be appropriate.  

These methods are currently implemented in Tpetra.  We'd rather not have the hard-coded dependence on cusolve, however, as interfaces to on-node TPLs are KokkosKernels' responsibility.

**Comment by e10harvey** [comment]
@srajama1: I see your comment [here](https://github.com/trilinos/Trilinos/pull/6813?email_source=notifications&email_token=ANUVLYQAOVYOPRW5RJRGMVLRCQIF5A5CNFSM4KTFPAL2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCVH2VCA#pullrequestreview-357542536). Do you want me to look into this issue as well?

**Comment by srajama1** [comment]
@e10harvey : Yes, this is what I was talking about need for several kernels. Start with the three kernels @iyamazaki wanted as they are needed for an application in short time. This one already went in, we can clean it up as the next step.

**Comment by lucbv** [comment]
With the merge of PR #3056 we have implemented the three QR algorithms: qrf, mqr and gqr for CPUs, CUDA and HIP.
Closing this as completed!
